fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+                    ingest_summary_node_mappings.push(IngestSummaryNodeMapping {
+                        node_index: ingest_client_index,
+                        ingest_summary,
+                    });
+                }
+                Err(_) => {
+                    log::warn!(
+                        self.logger,
+                        "Unable to retrieve ingest summary for node: {}",
+                        ingest_client.get_uri()
+                    );
+                }
+            }
+        }
+
+        ingest_summary_node_mappings
+    }
+
+    /// Performs automatic failover, which means that we try to activate nodes
+    /// for an outstanding ingress key, if it exists.
+    ///
+    /// The logic is as follows:
+    ///   1. Find all of the "outstanding keys" as determined by the RecoveryDb.
+    ///      These are ingress keys that Fog Ingest needs to scan blocks with
+    ///      but Fog Ingest isn't currently doing that because all nodes are
+    ///      idle.
+    ///   2. If there are:
+    ///         a) 0 outstanding keys:
+    ///              No node will be activated. We now have to:
+    ///                 (i)  Set new keys on an idle node.
+    ///                 (ii) Activate that node.
+    ///         b) 1 oustanding key:
+    ///              Try to find an idle node that contains that key.
+    ///                 (i)  If you find one, great! Just activate that node. If
+    ///                      activation is unsuccessful, then return an error
+    ///                      and return to the overseer polling logic.
+    ///                 (ii) If you don't find an idle node with that key,
+    ///                      then you have to report that key as lost.
+    ///        c) > 1 outstanding key:
+    ///             (i) Disable
+    ///             (ii) TODO: Send an alert.
+    fn perform_automatic_failover(
+        &self,
+        ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping>,
+    ) -> Result<(), OverseerError> {
+        let inactive_outstanding_keys: Vec<CompressedRistrettoPublic> =
+            self.get_inactive_outstanding_keys()?;
+
+        match inactive_outstanding_keys.len() {
+            0 => log::info!(self.logger, "Found 0 outstanding keys."),
+            1 => log::info!(self.logger, "Found 1 outstanding key."),
+            _ => {
+                log::error!(self.logger, "Found multiple outstanding keys. This requires manual intervention. Disabling overseer.");
+                self.is_enabled.store(false, Ordering::SeqCst);
+                return Err(OverseerError::MultipleOutstandingKeys("Multiple outstanding keys found. This is unexpected and requires manual intervention. As such, we've disabled overseer. Take action and then enable overseer.".to_string()));
+            }
+        }
+
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
+            match self.try_to_activate_node_for_inactive_outstanding_key(
```

Comment:
I think this `try_to_activate_node` function is a little too busy -- there are errors that can occur if contacting a node fails, and errors that can occur if finding the key fails it seems. This seems bad for clear error handling.

What I would suggest is that we try to move more logic from that function to this function, and possibly eliminate that function:

* You could iterate across the `ingest_summary_node_mappings`, using `iter().find()`, or a for loop, whatever you think is more clear
* Try to activate the first node that has the desired key. If that fails, return an error, and we'll re-evaluate everything and try again in the next pass
* If no node has the desired key, then we report a lost key, and try to activate the first node

In general any errors that we get when talking to ingest should not be just "log and fall through", they should cause us to abort, and return an error to the caller, who logs it. This will make it much easier to figure out in what circumstances the overseer may attempt to activate something. We want that circumstance to be as simple and as restrictive as possible. It should only do this when we are as certain as we can be that this is the right thing to do.

Reply:
Agreed. Thank you for expressing this so clearly. Implemented it as you described.